sentry_streams/src/broadcaster.rs

```diff
@@ -0,0 +1,262 @@
+            pending_messages: HashMap::new(),
+        }
+    }
+
+    /// Attempts to re-submit a pending message, if successful deletes it from the pending buffer.
+    fn retry_pending_message(
+        &mut self,
+        message: Message<RoutedValue>,
+        identifier: &MessageIdentifier,
+    ) -> Result<(), SubmitError<RoutedValue>> {
+        self.next_step.submit(message)?;
+        self.pending_messages.remove(identifier);
+        Ok(())
+    }
+
+    /// Attempts to re-submit all pending messages.
+    fn flush_pending(&mut self) -> Result<(), StrategyError> {
+        let ids: Vec<MessageIdentifier> = self.pending_messages.keys().cloned().collect();
+        for identifier in ids {
+            let msg = self.pending_messages.get(&identifier).unwrap();
+            // we only need to take action here if the returned error is `InvalidMessage`
+            if let Err(SubmitError::InvalidMessage(e)) =
+                self.retry_pending_message(msg.clone(), &identifier)
+            {
+                self.pending_messages.remove(&identifier);
+                return Err(e.into());
+            }
+        }
+        Ok(())
+    }
+
+    /// Attempts to submit a message to the next step, if backpressure
+    /// is raised then adds the message to the pending buffer.
+    fn submit_to_next_step(
+        &mut self,
+        message: Message<RoutedValue>,
+        identifier: MessageIdentifier,
+    ) -> Result<(), SubmitError<RoutedValue>> {
+        let msg_clone = message.clone();
+        match self.next_step.submit(message) {
+            Ok(..) => Ok(()),
+            Err(e) => {
+                self.pending_messages.insert(identifier, msg_clone);
+                Err(e)
+            }
+        }
+    }
+
+    // If the message is in the pending buffer, attempts to submit it to the
+    /// next step
+    fn handle_submit(
+        &mut self,
+        message: Message<RoutedValue>,
+    ) -> Result<(), SubmitError<RoutedValue>> {
+        let identifier = MessageIdentifier {
+            route: message.payload().route.clone(),
+            committable: clone_committable(&message),
+        };
+        if self.pending_messages.contains_key(&identifier) {
+            self.retry_pending_message(message, &identifier)
+        } else {
+            self.submit_to_next_step(message, identifier)
+        }
+    }
+}
+
+impl ProcessingStrategy<RoutedValue> for Broadcaster {
+    fn poll(&mut self) -> Result<Option<CommitRequest>, StrategyError> {
+        self.flush_pending()?;
+        self.next_step.poll()
+    }
+
+    fn submit(&mut self, message: Message<RoutedValue>) -> Result<(), SubmitError<RoutedValue>> {
+        if self.route != message.payload().route {
+            return self.next_step.submit(message);
+        }
+        let unfolded_messages = generate_broadcast_messages(&self.downstream_branches, message);
+        for msg in unfolded_messages {
+            self.handle_submit(msg)?;
+        }
```

Comment:
I don't think this works as expected.
As this piece of code does not flush the pending messages, if there are pending messages, you would submit the newer message before the pending messages (that are older) thus sending messages out of order, which is wrong.

Please ensure this case is covered by your unit tests.